Improve compile error for borrowing mut while missing Sized trait bound

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.

If you cannot produce a minimal reproduction case (something that would work in
isolation), please provide the steps or even link to a repository that causes
the problematic output to occur.
-->


Given the following [code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=26d81fb56dbc4f01b3a3da064b737f7b): 

```rust
trait Modify {
    fn modify(&mut self) ;
}

impl<T> Modify for T  {
    fn modify(&mut self)  {
        // ...
    }
}

trait Foo {
    fn mute(&mut self) {
        self.modify();
    }
}
```

The current output is:

```
self.modify();
^^^^^^^^^^^^^ cannot borrow as mutable
```

<!-- The following is not always necessary. -->
Ideally the output should look like:

```
the size for values of type `Self` cannot be known at compilation time
```

add `Sized` trait bound and the code compiles successfully:

```rust
trait Modify {
    fn modify(&mut self) ;
}

impl<T> Modify for T  {
    fn modify(&mut self)  {
        // ...
    }
}

trait Foo: Sized {
    fn mute(&mut self) {
        self.modify();
    }
}
```

rust version:  rustc 1.58.0 (02072b482 2022-01-11)

[This came up on URLO in the context of an extension trait,](https://users.rust-lang.org/t/cannot-borrow-self-as-mutable-when-self-is-mut-self/83277) and I'd like to point out how unhelpful the error is:
* Given that the signature is `fn mute(&mut self) {`, it's confusing to  say
   ```rust
   error[E0596]: cannot borrow `self` as mutable, as it is not declared as mutable
     --> src/lib.rs:13:9
      |
   13 |         self.modify();
      |         ^^^^^^^^^^^^^ cannot borrow as mutable
   ```
* This is the only thing that is said, there is no suggestion
* You have to figure out it means the `&mut Self` is not mutable, vs. the referenced `Self`
* You have to know about method resolution to have a chance of understanding it's trying to create a `&mut &mut self` to call the implementation of `&mut Self`
* You have to know about implicit `Sized` bounds and where they *aren't* to have a chance of understanding *why* it's not calling an implementation of `Self`

It is a bit tricky though, as there are at least three possible suggestions:
* Change the signature to `fn mute(mut self: &mut Self) {`
* Add a `Sized` bound to `trait Foo`
* Add a `?Sized` bound on the blanket implementation:
   ```rust
   impl<T: ?Sized> Modify for T  {
   ```

The last solution is the most correct in my opinion (but can only be applied if that implementation is local).

Finally, [this variation:](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3b51aa30caac2bb9403599296851f0d3)
```rust
    fn mute(self: &mut Self) {
        self.modify();
    }
```
*Does* give a suggestion:
```rust
help: consider making the binding mutable
   |
11 |     fn mute(mut self: &mut Self) {
   |             +++
```
Even though it's not the ideal suggestion.  Let me know if I should file a separate bug for `&mut self` sugar disabling diagnostic suggestions.  (If it exists I didn't find it.)

@rustbot label +D-confusing +D-terse +D-newcomer-roadblock

--- a/compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs
+++ b/compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs
@@ -344,20 +344,25 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
                     } else {
                         err.span_help(source_info.span, "try removing `&mut` here");
                     }
-                } else if decl.mutability == Mutability::Not
-                    && !matches!(
+                } else if decl.mutability == Mutability::Not {
+                    if matches!(
                         decl.local_info,
                         Some(box LocalInfo::User(ClearCrossCrate::Set(BindingForm::ImplicitSelf(
                             hir::ImplicitSelfKind::MutRef
-                        ))))
-                    )
-                {
-                    err.span_suggestion_verbose(
-                        decl.source_info.span.shrink_to_lo(),
-                        "consider making the binding mutable",
-                        "mut ",
-                        Applicability::MachineApplicable,
-                    );
+                        ),)))
+                    ) {
+                        err.note(
+                            "as `Self` may be unsized, this call attempts to take `&mut &mut self`",
+                        );
+                        err.note("however, `&mut self` expands to `self: &mut Self`, therefore `self` cannot be borrowed mutably");
+                    } else {
+                        err.span_suggestion_verbose(
+                            decl.source_info.span.shrink_to_lo(),
+                            "consider making the binding mutable",
+                            "mut ",
+                            Applicability::MachineApplicable,
+                        );
+                    };
                 }
             }
 

--- a/src/test/ui/borrowck/issue-93078.rs
+++ b/src/test/ui/borrowck/issue-93078.rs
@@ -1,0 +1,15 @@
+trait Modify {
+    fn modify(&mut self) ;
+}
+
+impl<T> Modify for T  {
+    fn modify(&mut self)  {}
+}
+
+trait Foo {
+    fn mute(&mut self) {
+        self.modify(); //~ ERROR cannot borrow `self` as mutable
+    }
+}
+
+fn main() {}

--- a/src/test/ui/borrowck/issue-93078.stderr
+++ b/src/test/ui/borrowck/issue-93078.stderr
@@ -1,0 +1,12 @@
+error[E0596]: cannot borrow `self` as mutable, as it is not declared as mutable
+  --> $DIR/issue-93078.rs:11:9
+   |
+LL |         self.modify();
+   |         ^^^^^^^^^^^^^ cannot borrow as mutable
+   |
+   = note: as `Self` may be unsized, this call attempts to take `&mut &mut self`
+   = note: however, `&mut self` expands to `self: &mut Self`, therefore `self` cannot be borrowed mutably
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0596`.

